gBoards GergoPlex

pirj opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->

## Description

Add firmware for [gBoards GergoPlex](https://www.gboards.ca/product/gergoplex).

## Not done

 - `SPLIT_KEYBOARD=yes` could potentially reduce the amount of custom code
 - in the custom matrix code rows and cols seem to be mixed up. Might be it's intentional
 - the code is reading directly using `PINF` (inherited from #8928)

## Types of Changes

- [x] Keyboard (addition or update)

## Issues Fixed or Closed by This PR

fixes #8262, #8928, #10406

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

btannous commented:
looks like everything works except the mouse functionality, neither clicking or moving seem to work (which you mentioned the lmb and rmb do not seem to work)

btannous commented on keyboards/gboards/gergoplex/keymaps/default/rules.mk:
```diff
@@ -0,0 +1,19 @@
+# Make sure you have dfu-programmer installed!
+#----------------------------------------------------------------------------
+# Firmware options
+MOUSEKEY_ENABLE        = no
```
This line to `yes` might fix the mouse issues.

pirj replied:
Oh 🙈 
Thank you for spotting this!

btannous commented:
https://github.com/pirj/qmk_firmware/pull/1

updated the delay:
 - it was set to 40
 - docs seem to show 300
 - so went with 400
 - https://beta.docs.qmk.fm/using-qmk/software-features/feature_combo

also the `mousekey_enable = no` in the keymap seemed to override whats expected, so changed that to yes.

seems much more usable with 400 vs 40!

pirj commented:
@btannous Incorporated your fixes with an adjustment - removed the redundant `MOUSEKEY_ENABLE` from `keymaps/default/rules.mk` so it falls back to `rules.mk`; removed the `COMBO_TERM` so it falls back to `TAPPING_TERM` that is 200 by default.
Thank you!

btannous commented:
@pirj awesome work.  `IGNORE_MOD_TAP_INTERRUPT` seems to fix my issue with the CTRL/Shift modifiers, so sent you a PR for that one line.

I'm using your latest branch + that config change and so far this seems to work well.

pirj commented:
@btannous Thanks, merged!
I am sorry for thoughtlessly removing this option.

btannous commented:
All good, that change makes a huge difference on those keys.  Now Im hopeful this is good to the QMK maintainers!  Thanks all :)

btannous commented:
Also +@germ

drashna commented on keyboards/gboards/gergoplex/gergoplex.c:
```diff
@@ -0,0 +1,52 @@
+#include "gergoplex.h"
```
Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

drashna commented on keyboards/gboards/gergoplex/gergoplex.c:
```diff
@@ -0,0 +1,52 @@
+bool i2c_initialized = 0;
+i2c_status_t mcp23018_status = 0x20;
+
+void matrix_init_kb(void) {
+    matrix_init_user();
+}
```
```suggestion
```

drashna commented on keyboards/gboards/gergoplex/gergoplex.h:
```diff
@@ -0,0 +1,40 @@
+#pragma once
```
Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

drashna commented on keyboards/gboards/gergoplex/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,97 @@
+ * https://github.com/qmk/qmk_firmware/blob/master/docs/keycodes.md
+ */
+
+#include "gergoplex.h"
```
```suggestion
#include QMK_KEYBOARD_H
```

drashna commented on keyboards/gboards/gergoplex/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,97 @@
+/* Good on you for modifying your layout! if you don't have
```
Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

drashna commented on keyboards/gboards/gergoplex/matrix.c:
```diff
@@ -0,0 +1,241 @@
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "matrix.h"
```
This should be converted to matrix lite.

Check the ergodox ez's matrix for reference.

pirj replied:
Done so in the last commit, but the keyboard just stopped working.
Can you please give a cue what did I do wrong, @drashna ?

drashna replied:
Hmm.  My guess is that some of the config isn't correct. 

It's just the expander side that stopped working?

btannous replied:
In `static matrix_row_t read_cols(uint8_t row) {` @germ originally did a write before reading i2c, it looks like that's missing, maybe thats it? (I'm new to qmk_firmware, so not sure)

https://github.com/germ/qmk_firmware/blob/39c86e080dc04b68ee08387dcb5144c88cb44855/keyboards/gboards/k/gergoplex/matrix.c#L223-L224

pirj replied:
> It's just the expander side that stopped working?

Both, actually. Is this a clue?

drashna replied:
I'm sorry, I lost track of this. 

And yeah, it means that the matrix isn't being processed correctly.  Give me a couple of days, and I'll see if I can fix up some code. 

Unfortunately, I don't have a gergoplex to test on, and I'm not sure how close the other boards are in design, so testing is ... problematic.  

If it's more than a few days, give me a ping on the discord.

pirj replied:
Thanks. I might have some time this weekend to debug it. On the one hand (pun intended) it's confusing that both parts of the keyboard stop working, even though one of the matrices is scanned via I2C, and another with matrix lite's helper code.
The other thing that confuses me is that rows and columns are mixed up.
A good thing is that if you checkout `@~1`, you have a working code.

drashna replied:
Yeah, that design is pulled from the ergodox (og), and ... IIRC it's to help with scan speeds

drashna commented on keyboards/gboards/gergoplex/rules.mk:
```diff
@@ -0,0 +1,18 @@
+
+BOOTLOADER = atmel-dfu
+
+CUSTOM_MATRIX    = yes
```
```suggestion
CUSTOM_MATRIX    = lite
```

George-Lee commented:
Is there anything I can do as an amateur with C to help with this? (waiting on my gergoplex to be shipped so no device yet)

stjoni commented:
I have not attempted to debug, but merging manna-harbour/miryoku into this branch and `make gboards/gergoplex:manna-harbour_miryoku` builds without errors, but the keyboard is not functional after flashing.

caksoylar commented:
@stjoni I believe the latest commit in the PR is still WIP with the custom matrix lite refactor. Commits up to and including https://github.com/qmk/qmk_firmware/pull/13027/commits/06b85ec0376fe5598ac98ff271167a484044208b builds and works fine on my end.

drashna commented:
For now, it may be simplest/best to revert the matrix lite code.

pirj commented:
@drashna Thank you! Will handle it over the weekend.
It was a slight improvement code-wise, but not as significant as I imagined it would be to reuse the commonly used code.

pirj commented:
@drashna I've reset the commit with matrix lite.

caksoylar commented on keyboards/gboards/gergoplex/rules.mk:
```diff
@@ -0,0 +1,18 @@
+
+LAYOUTS = split_3x5_3
+
+DEBOUNCE_TYPE  = eager_pr
```
```suggestion
DEBOUNCE_TYPE  = sym_eager_pr
```
This fails building with current `develop` since the old names seem to be removed.

pirj replied:
Thanks!

pirj commented:
Rebased on `develop`.

pirj commented:
Something's not right. I've reinstalled QMK package in my system, and rebased the PR on `develop`. Now the previously working (pre matrix lite) firmware doesn't work. It builds, flashes, but the keyboard does not react.
I've checked out my `prettify-gergoplex-with-matrix-lite` that isn't rebased onto recent `develop`, checked out `HEAD~1` to remove matrix lite changes, the same issue.
I've checked out Germ's `gboards-gergoplex` (#8928), but it doesn't even build.

My suspicion is that it's me removing duplicated method definitions due to the newly appeared linker error:
```
quantum/matrix_common.c:36: multiple definition of `matrix_is_on'; .build/obj_gboards_gergoplex_default/matrix.o:/Users/pirj/source/qmk_firmware/keyboards/gboards/gergoplex/matrix.c:160: first defined here
```
I had to remove those methods for the build to work.

However, a warning has caught my eye yesterday, something like `MATRIX_ROWS doesn't match MATRIX_ROWS_PINS, 10 vs 4` and a similar related to columns.
Keeping in mind that rows and cols are horribly mixed up here:
```
// Remember this means ROWS
static void init_cols(void) {
    for (uint8_t row = 0; row < MATRIX_COLS; row++) {
```
```
static const pin_t row_pins[MATRIX_COLS] = MATRIX_ROW_PINS;
// Right-hand side only pins, the left side is controlled my MCP
static const pin_t col_pins[MATRIX_ROWS_PER_SIDE] = MATRIX_COL_PINS;
```
I suspect that the definition of `matrix_key_count` from `quantum/matrix.c`:
```
uint8_t matrix_key_count(void) {
    uint8_t count = 0;
    for (uint8_t i = 0; i < MATRIX_ROWS; i++) {
        count += matrix_bitpop(i);
    }
    return count;
}

#if (MATRIX_COLS <= 8)
#    define matrix_bitpop(row) bitpop(matrix_get_row(row))
#elif (MATRIX_COLS <= 16)
#    define matrix_bitpop(row) bitpop16(matrix_get_row(row))
```

doesn't do good job replacing our definition (`keyboards/gboards/gergoplex/matrix.c`) when provided an incorrect `MATRIX_COLS`:
```
uint8_t matrix_key_count(void) {
    uint8_t count = 0;
    for (uint8_t i = 0; i < MATRIX_ROWS; i++) {
        count += bitpop16(matrix[i]);
    }
    return count;
}
```

Thus is due to `MATRIX_COLS` being set to 4 instead of 10. And, as a result, `bitpop` used instead of `bitpop16`.

I tried to confirm this theory by temporarily patching `quantum/matrix_common.c`:
```diff
#if (MATRIX_COLS <= 8)
- #    define matrix_bitpop(row) bitpop(matrix_get_row(row))
+ #    define matrix_bitpop(row) bitpop16(matrix_get_row(row))
```

but the firmware still didn't work.

The last working firmware is at https://github.com/btannous/qmk_firmware/tree/prettify-gergoplex

I must have broken something after that.

pirj commented:
Silly me, I forgot to remove two commits related to custom matrix lite. Removed. Now everything works fine. 🎉 

@btannous I've incorporated your [changes for Colemak Mod-DH](https://github.com/pirj/qmk_firmware/pull/3). I've kept `germ-mouse-keys.def` separate, just in case someone would want to exclude them, and it's as easy as editing their `combos.def`. Thanks!

manna-harbour commented on keyboards/gboards/gergoplex/keymaps/colemak-dhm/rules.mk:
```diff
@@ -0,0 +1,17 @@
+
+ifeq ($(strip $(DEBUG_MATRIX)), yes)
+    OPT_DEFS += -DDEBUG_MATRIX
+endif
```
Enable [community layout](https://beta.docs.qmk.fm/developing-qmk/qmk-reference/feature_layouts) support:

```suggestion
endif

LAYOUTS = split_3x5_3
```

George-Lee commented:
Is there anything we/I can do to help with this?

btannous commented:
![YAY](http://www.reactiongifs.com/r/swih.gif)

btannous commented:
Typing this on my georgoplex after I checked out https://github.com/pirj/qmk_firmware/tree/prettify-gergoplex and merged with a freshly synced master then flashed with qmk on an m1 chip. all works well. cant wait till this finally is going to get merged in.  thanks everyone!

pirj commented:
Thank you, **@germ**, @btannous, @manna-harbour, **dear maintainers**, and early testers @YorickPeterse, @stjoni and @caksoylar for this to happen! 🎉
